hobbify/habits/views/habits.py

```diff
@@ -0,0 +1,209 @@
+from django_filters.rest_framework import DjangoFilterBackend
+
+# Serializers
+from hobbify.habits.serializers import HabitModelSerializer
+from hobbify.habits.serializers import RecurrenceModelSerializer
+
+# Dateutil
+from dateutil.rrule import rrule, DAILY, MONTHLY, WEEKLY
+
+# Models
+from hobbify.habits.models import Habit
+from hobbify.users.models import User
+from hobbify.habits.models import RecurrentInstance
+
+# Utils
+import datetime
+
+
+class HabitViewSet(mixins.ListModelMixin,
+                   mixins.CreateModelMixin,
+                   mixins.RetrieveModelMixin,
+                   mixins.UpdateModelMixin,
+                   viewsets.GenericViewSet):
+    """ Habit view set """
+
+    lookup_field = 'id'
+    serializer_class = HabitModelSerializer
+
+    filter_backends = (SearchFilter, OrderingFilter, DjangoFilterBackend)
+    ordering = ('created', 'end_date', 'start_date')
+    ordering_fields = ('created', 'end_date', 'modified', 'start_date')
+    search_fields = ('name', 'description')
+    filter_fields = ('paused', 'is_private')
+
+    def dispatch(self, request, *args, **kwargs):
+        """ Verify that the user exists """
+        username = kwargs['username']
+        self.user = get_object_or_404(User, username=username)
+        return super(HabitViewSet, self).dispatch(request, *args, **kwargs)
+
+    def get_serializer_context(self):
+        """ Add user to serializer context """
+        context = super(HabitViewSet, self).get_serializer_context()
+        context['user'] = self.user
+        return context
+
+    def get_queryset(self):
+        """ Return not hidden habits """
+        queryset = Habit.objects.filter(is_hidden=False)
+        if self.action == 'list':
+            username = self.kwargs['username']
+            return Habit.objects.filter(is_hidden=False, owner__username=username)
+        return queryset
+
+    def get_permissions(self):
+        """ Assign permission based on action """
+        permissions = [IsAuthenticated,]
+        return [p() for p in permissions]
+
+    @action(methods=['get', 'patch'], detail=True)
+    def instance(self, request, *args, **kwargs):
+        """ Return and partial update instance by id """
+        instance = RecurrentInstance.objects.filter(habit=self.kwargs['id'], is_hidden=False, id=self.kwargs['r_id'])
+        if request.method == 'PATCH':
+            instance.done = request.data.done
+        return Response(instance, status=status.HTTP_200_OK)
+
+    @action(methods=['get'], detail=True)
+    def recurrences(self, request, *args, **kwargs):
+        """ Return the past recurrences of an habit """
+        recurrences = RecurrentInstance.objects.filter(habit=self.kwargs['id'], is_hidden=False)
+        data = [RecurrenceModelSerializer(r).data for r in recurrences]
+        return Response(data, status=status.HTTP_200_OK)
+
+    @action(detail=True, methods=['post'])
+    def recurrence(self, request, *args, **kwargs):
```

Comment:
I wouldn't expect the users to create a recurrence. It should be the backend's job to know when to create them when listing.